File: ProjectPO_Java/src/PO_java/LifeSymulator/Plants/SosnowskyHogweed.java
package PO_java.LifeSymulator.Plants;

import PO_java.LifeSymulator.Animal;
import PO_java.LifeSymulator.Organism;
import PO_java.LifeSymulator.Plant;
import PO_java.LifeSymulator.Tools.Cords;
import PO_java.LifeSymulator.World;

import java.util.Random;

public class SosnowskyHogweed extends Plant {
    private final int POSSIBLEMOVES = 4;
    public SosnowskyHogweed(World world, Cords cords)
    {
        strength = 10;
        age = 0;
        sign = '#';
        name = "SosnowskyHogweed";
        toKill = false;
        isDead = false;
        this.cords = cords;
        this.world = world;
    }

    @Override
    public void action(){
        Random rand = new Random();
        int chance = rand.nextInt(100/PROBABILITY);
        if(chance == LUCKY){
            Cords seedCords = new Cords(-1, -1);
            if (checkFreeFieldsNear(this.cords))
                seedCords.setCords(randomFreeFieldsNear(this.cords));
            if((seedCords.getX() != -1 && (seedCords.getY() != -1)))
            {
                Organism tmp = world.createOrganism(this.name, seedCords);
                world.addOrganismToBoard(tmp);
                world.addOrganismToList(tmp);
                System.out.println("New " + tmp.getName() + " seed was sown on field ("
                        + seedCords.getX() + ", " + seedCords.getY() + ")");
            }
        }
        killingNear();
    }

    @Override
    public void collision(Organism occupyingOrganism){
        world.addOrganismToKill(this);
        world.addOrganismToKill(occupyingOrganism);
        world.cleanBoardField(this.cords);
        world.cleanBoardField(occupyingOrganism.getCords());
        System.out.println(this.name + " killed " + occupyingOrganism.getName() + " while they was trying to eat it");
    }

    public void killingNear(){
        int x = cords.getX();
        int y = cords.getY();
        for (int i = 0; i < POSSIBLEMOVES; i++)
        {
            Cords possMove = movePossibilities(cords, i);
            if (checkMove(possMove) && world.isSomethingThere(possMove))
            {

                if (world.whatIsThere(possMove) instanceof Animal)
                {
                    Organism organism = world.whatIsThere(possMove);
                    world.addOrganismToKill(organism);
                    world.cleanBoardField(organism.getCords());
                    if(world.getOrganismsBoard()[possMove.getY()][possMove.getY()] == null)
                        System.out.println(organism.getName() + " was killed by SosnowskyHogweed!!!!!!!!!!!!!!! " + possMove.getX() + ", " + possMove.getY());
                }
            }
        }
    }

}

File: ProjectCppPO/SosnowskyHogweed.cpp
#include "SosnowskyHogweed.h"
#define POSSIBLEMOVES 4

SosnowskyHogweed::SosnowskyHogweed(World* world, Cords cords)
{
	features.strength = 10;
	features.age = 0;
	features.look = '#';
	features.name = "Sosnowsky Hogweed";
	features.toKill = false;
	location = cords;
	this->world = world;
}

//Metoda nadpisuj¹ca metodê z klasy rodzica ¿eby barszcz sosnowskiego zabija³ ka¿dego kto stoi obok niego
void SosnowskyHogweed::action()
{
	int chance = rand() % (100 / PROBABILITY);
	if (!chance)
	{
		Cords seedCords(-1, -1);
		if (checkFreeFieldsNear(this->location))
			seedCords.setCords(randomFreeFieldsNear(this->location));
		if ((seedCords.getX() != -1) && (seedCords.getY() != -1))
		{
			Organism* tmpOrganism = world->createOrganism(features.name, seedCords);
			world->addOrganismToBoard(tmpOrganism);
			world->addOrganismToList(tmpOrganism);
			string report = "New " + tmpOrganism->getFeatures().name + " seed was sown on field ("
				+ to_string(seedCords.getX()) + ", " + to_string(seedCords.getY()) + ")";
			world->addReport(report);
		}
	}
	killingNear();
}

//Metoda nadpisuj¹ca metodê z klasy rodzica ¿eby marszcz zabija³ ka¿de zbierze które go zje
void SosnowskyHogweed::collision(Organism* occupyingOrganism)
{
	world->addOrganismToKill(this);
	world->addOrganismToKill(occupyingOrganism);
	world->cleanBoardField(location);
	world->cleanBoardField(occupyingOrganism->getLocation());
	//dodanie raportu
	string report = features.name + " killed " + occupyingOrganism->getFeatures().name + " while they was eating it";
	world->addReport(report);
}

//Metoda obs³uguj¹ca zabijanie pobliskich organizmów przez barszcz
void SosnowskyHogweed::killingNear()
{
	int x = location.getX();
	int y = location.getY();
	for (int i = 0; i < POSSIBLEMOVES; i++)
	{
		if (checkMove(movePossibilities(location, i)) && world->isSomethingThere(movePossibilities(location, i)))
		{
			Organism* organism = world->whatIsThere(movePossibilities(location, i));
			if (organism->isItAnimal())
			{
				world->addOrganismToKill(organism);
				world->cleanBoardField(organism->getLocation());
				//dodanie raportu
				string report = features.name + " killed " + organism->getFeatures().name;
				world->addReport(report);
			}
		}
	}
}
